Add exportCSV to dt-table

LucasHocker opened the pull request:
### <strong>Pull Request</strong>

<hr>
Hi, thank you for contributing to Barista with this pull request (PR).

To ensure a fast process and merging of your PR please make sure it fulfills the
coding standards and contribution guidelines.

- A feature proposal has been provided, discussed and approved first.
- There is a meaningful description of the issue in GitHub (Screenshots are
  often helpful).
- If the PR introduces breaking-changes or deprecations it matches the following
  guidelines.
  - The commit message follows our commit guidelines.
  - Tests for the changes have been added (for bug fixes / features).
  - Docs have been added / updated (for bug fixes / features).

Please choose the type appropriate for the changes below: <br>

#### Type of PR

<!-- Bugfix (non-breaking change which fixes an issue) -->
<!-- Feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or change that would cause existing functionality to not work as expected) -->
<!-- Documentation update (changes to documentation) -->
<!-- Other (if none of the above apply) -->
Feature: added a context dialogue with an "Export CSV" button

#### Checklist

- [x] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [x] Lint and unit tests pass locally with my changes
- [n/a] I have added tests that prove my fix is effective or that my feature works
- [n/a] I have added necessary documentation (if appropriate)

tomheller reviewed (changes requested):
Hi @LucasHocker! Thank you for your contribution.

This looks very good so far. A couple of things to note here, please make the display / usage of the context dialog an `opt-in`. The way you have implemented this now, it will appear on every barista table, without the consumer having the ability to influence this behavior.

With an opt-in switch you can make sure that there is also enough space on the top of the table to position the context-dialog correctly. The negative top and margin values are usually not something we work with, because it can create unpredictable outcome, depending on where on the page the table is rendered. Here is an example from our demos application, that shows that this might be a bit broken.
![image](https://user-images.githubusercontent.com/6338434/142982843-c7aec125-df9a-4bb4-a94f-a01e9623b410.png)

Additionally, please make sure that you revert the changes from the `package-lock.json`, they should not be part of this PR.

LucasHocker commented:
Thanks Thomas. I suspected there might be a couple concerns with the mandatory approach. Would you be open to talking about an opt-out approach versus opt-in? Relying on each page owner to go back and opt-in could have very low adoption rates due to various other priorities. As an end user, I want to be able to export tabular data any time I encounter it. I don’t think that could be achieved with an opt-in. Thoughts?

I’ll work on a better way to position the context dialogue rather than relative positioning. Most examples of dt-table seem to have white space above the right side of the table, so I was hoping to utilize that rather than creating another row of white space. I’ll play around with some more examples.

The package.json was unintentional; I’ll be sure to remove that change.

LucasHocker commented:
Made the following changes:
- Moved export button to bottom, aligns with pagination row if present.
- Added opt-out Input.
- Added demo showing opt-out.
- Updated doc to explain opt-out.

Example (w/ pagination):
![Screen Shot 2021-11-23 at 11 30 03 AM](https://user-images.githubusercontent.com/34926264/143079887-dab921bd-9d29-4c96-8b21-63f6ba78d6e8.png)

Example (w/o pagination):
![Screen Shot 2021-11-23 at 11 31 46 AM](https://user-images.githubusercontent.com/34926264/143080074-5f036c55-c69c-416e-b5d3-4b123a8a1ccc.png)

Example (opt-out):
![Screen Shot 2021-11-23 at 12 06 56 PM](https://user-images.githubusercontent.com/34926264/143080183-b79bae47-b281-4951-919b-bafe369e7664.png)

LucasHocker commented:
Eventually got all the CI tests to succeed, but `pull_request_labeler` step seems to be stuck. Hopefully you can bump that when you get a change to review the change.

tomheller commented:
> Eventually got all the CI tests to succeed, but pull_request_labeler step seems to be stuck. Hopefully you can bump that when you get a change to review the change.

This is design to only run if certian finishing labels are being applied. It's fine if it is stuck for now. 

One more observation, can you please switch the `opt-out` switch to an `opt-in` one as outlined in my comment above. We do not want to touch every table in the product with an update like this.

LucasHocker commented:
Updated (hopefully) ready to merge

tomheller commented on libs/barista-components/table/src/table.ts:
```diff
@@ -62,6 +62,7 @@ import {
   _COALESCED_STYLE_SCHEDULER,
 } from '@angular/cdk/table';
 import { ViewportRuler } from '@angular/cdk/scrolling';
+import { DtTableDataSource } from '..';
```
Please do not import from a barrel file here. Will fix that for you during cleanup, but this would lead to cyclic dependencies as this files is also exported in the barrel.

tomheller commented on libs/barista-components/table/src/table.ts:
```diff
@@ -141,6 +143,15 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
 
+  /** Whether the export button should be displayed. */
+  @Input()
+  get exportButton(): boolean {
```
```suggestion
  get showExportButton(): boolean {
```

tomheller commented on libs/barista-components/table/src/table.ts:
```diff
@@ -321,4 +332,41 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
   /** CSS class added to any row or cell that has sticky positioning applied. */
   protected stickyCssClass = 'dt-table-sticky';
+
+  exportCSV(): void {
```
Missing JSDoc, also this can be considered an internal function. Would need to mark it as such in the JSDoc and prefix with an _

```suggestion
/** @internal Converts the datasources data to csv format, adds it to a blob and initiates a download
  _exportCSV(): void {
```

tomheller commented on libs/barista-components/table/src/table.ts:
```diff
@@ -321,4 +332,41 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
   protected stickyCssClass = 'dt-table-sticky';
+
+  exportCSV(): void {
+    let ds = this.dataSource as DtTableDataSource<T>;
```
Most of the variable declarations here could be `const. We try to use constants wherever possible

tomheller commented on libs/barista-components/table/src/table.ts:
```diff
@@ -321,4 +332,41 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
+    if (typeof exportData[0] == 'object') {
+      keys = Object.keys(exportData[0]);
+    }
+    if (!keys.length) return;
```
Codestyle
```suggestion
    if (!keys.length) {
      return;
    }
```

tomheller commented on libs/barista-components/table/src/table.ts:
```diff
@@ -321,4 +332,41 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
+    if (!keys.length) return;
+    csv += keys.join(',') + '\n';
+
+    exportData.forEach((row) => {
```
Codestyle: prefer a for ... of loop when possible.

tomheller commented on libs/barista-components/table/src/table.ts:
```diff
@@ -321,4 +332,41 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
+      let vals = Object.values(row);
+      keys.forEach((key, idx) => {
+        let val = row[key];
+        if (typeof val != 'string') val = val.toString();
```
@LucasHocker The datasource can also hold complex values, not only primitives like the ones in your example. Data structures like this are not uncommon. 
```
 {
      host: 'et-demo-2-win4',
      cpu: 30,
      memory: {
        memoryPerc: 38,
        memoryTotal: 5830000000,
      },
      traffic: 98700000,
    },
  ...
  ```

With your implementation this would very likely result in a [Object object] stringification. I think we need another iteration on this part of the logic to get this rendering correctly.

tomheller commented on libs/barista-components/table/src/table.ts:
```diff
@@ -321,4 +332,41 @@ export class DtTable<T> extends _DtTableBase<T> implements OnDestroy {
+    });
+
+    //make csv document
+    let o = new Blob([csv], { type: 'text/csv' });
```
We might want to split this into a utility function. It does not necessarily be part of the member.

LucasHocker commented:
@tomheller Fixed above requested changes. Also did some refactoring after adding in selection specific export as well. Export functionality is now also documented in the Readme.md with live examples.
![Screen Shot 2021-12-07 at 10 46 41 AM](https://user-images.githubusercontent.com/34926264/145071123-343906be-eb4d-4492-a8d2-07321ed9097b.png)

Please take a look when you have a moment to see if we're aligned for a merge.

tomheller commented:
@LucasHocker Thank you for the effort. I will check it on Thursday, as we have a public holiday tomorrow in Austria.

tomheller commented:
@LucasHocker All looking good, I was trying to clean up some of the commits and code styles. You will need to give me permission to push to your fork (there should be a button somewhere that allows maintainers to push updates to your branch).

tomheller commented:
Cherry picked to release branch 10.x 9034a8be7aec7b7e43f3a50f711c561015da8d4f
